### PR TITLE
Introduce reroute method on IngestDocument

### LINKED ISSUE #83653: final_pipeline appears to process twice when using the date match index processor
### Elasticsearch Version

7.16.2

### Installed Plugins

_No response_

### Java Version

_bundled_

### OS Version

Cloud

### Problem Description

When using the `final_pipeline` option and the `date match index` in a pipeline to process data, the final_pipeline appears to run twice.  This can be shown by using the `append` processor which adds two of the same values or a `rename` processor which shows a failure, but it will show the name of a field changed and shows a failure due to the `old` named field not being present.

Also referenced in a [discuss topic](https://discuss.elastic.co/t/with-final-pipeline-and-a-pipeline-specified-the-final-pipeline-appears-to-be-running-twice-with-index-date-math-processor/296593) to confirm that the example is valid.

Additional note: This seems to run fine on a 7.6.2 environment, so something between that and 7.16.2 may have caused this issue.  #69727 or #75047 could be related.

### Steps to Reproduce

Append method of showing the `final_pipeline` running twice:

```
PUT _ingest/pipeline/test-final-pipeline
{
  "processors": [
    {
      "append": {
        "field": "pipeline",
        "value": "completed"
      }
    }
  ]
}

PUT _template/test-template
{
  "index_patterns": [
    "test*"
  ],
  "settings": {
    "final_pipeline": "test-final-pipeline",
    "number_of_shards": 2
  }
}

POST /test-1/_doc
{
  "time-field": "2021-02-10T12:02:01.789Z",
  "CONTENT": "blah1"
}

PUT /_ingest/pipeline/routing-test-pipeline
{
  "description": "Time series DAY pipeline",
  "processors": [
     {
         "date_index_name": {
             "field": "time-field",
             "index_name_prefix": "test-",
             "date_rounding": "d",
             "index_name_format": "yyyy-MM-dd"
         }
     }
 ]
}

POST /test-1/_doc?pipeline=routing-test-pipeline
{
  "time-field": "2021-02-10T12:02:01.789Z",
  "CONTENT": "blah2"
}
```

Rename processor used to show a failure.  
```
PUT _ingest/pipeline/test-final-pipeline
{
  "processors": [
    {
      "rename": {
        "field": "CONTENT",
        "target_field": "MALCONTENT"
      }
    }
  ]
}

PUT _template/test-template
{
  "index_patterns": [
    "test*"
  ],
  "settings": {
    "final_pipeline": "test-final-pipeline",
    "number_of_shards": 2
  }
}

POST /test-1/_doc
{
  "time-field": "2021-02-10T12:02:01.789Z",
  "CONTENT": "blah1"
}

PUT /_ingest/pipeline/routing-test-pipeline
{
  "description": "Time series DAY pipeline",
  "processors": [
     {
         "date_index_name": {
             "field": "time-field",
             "index_name_prefix": "test-",
             "date_rounding": "d",
             "index_name_format": "yyyy-MM-dd"
         }
     }
 ]
}

POST /test-1/_doc?pipeline=routing-test-pipeline
{
  "time-field": "2021-02-10T12:02:01.789Z",
  "CONTENT": "blah2"
}
```

### Logs (if relevant)

_No response_

### PR DESCRIPTION
**Pull request by felixbarny**
- Overrides _index
- Skips current pipeline
- Invokes default pipeline of new index

Requirement for
* https://github.com/elastic/elasticsearch/pull/76511

Combines https://github.com/elastic/elasticsearch/pull/85932 and https://github.com/elastic/elasticsearch/pull/85931

Fixes #83653

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/ingest/IngestService.java
Comment (dakrone) [inline server/src/main/java/org/elasticsearch/ingest/IngestService.java:779]: I think this is a reasonable abstraction, but I don't like that we make it mutable, and that we don't encapsulate enough here.

The `withoutDefaultPipeline` method makes me uncomfortable as its name makes it sound like it would return a new object rather than making a mutable change.

I also think we don't need to have `executePipelines(...)` take a boolean, we're treating this internally as though we'll always have a list of pipelines, but we could probably get away with passing a proper object instead `Iterator<String>`. I think it'd be much clearer that way.

File: server/src/main/java/org/elasticsearch/ingest/IngestDocument.java
Comment (dakrone) [inline server/src/main/java/org/elasticsearch/ingest/IngestDocument.java:925]: Why doesn't this also reset `invokeDefaultPipelineOfDestination`? Also, it's unclear from the code why we need this method to be invoked.

File: server/src/main/java/org/elasticsearch/ingest/IngestDocument.java
Comment (dakrone) [inline server/src/main/java/org/elasticsearch/ingest/IngestDocument.java:921]: I don't think we need multiple boolean flags for this right? If we do, then it's not clear enough why they have to be separated (needs documentation)

File: server/src/main/java/org/elasticsearch/ingest/IngestService.java
Comment (felixbarny) [reply server/src/main/java/org/elasticsearch/ingest/IngestService.java]: I abstracted the `Iterator<String>` and the boolean flag into a `PipelineIterator`, however it's not immutable. As that's the nature of iterators, I think that's fine.

The `executePipelines` needs to know about three properties of the current pipeline: the name (to add the name in the exception in case the pipeline itself can't be resolved), the pipeline itself, and whether the current pipeline is the final pipeline (if true, it's disallowed to override _index).

The `PipelineIterator` encapsulates that state and I think that makes it cleaner than before. Thanks for the suggestion 👍

File: server/src/main/java/org/elasticsearch/ingest/IngestDocument.java
Comment (felixbarny) [reply server/src/main/java/org/elasticsearch/ingest/IngestDocument.java]: Good point, it needs to reset both. I've changed it to just be a single boolean.

File: server/src/main/java/org/elasticsearch/ingest/IngestDocument.java
Comment (felixbarny) [reply server/src/main/java/org/elasticsearch/ingest/IngestDocument.java]: I've changed it to just be a single boolean and a single getter for it (`isRedirect()`). The downside is that `isRedirect()` is less expressive as `isSkipCurrentPipeline()` in the context of determining whether to skip the pipeline.
We could keep `isSkipCurrentPipeline()` and just return the `redirect` flag. But that feels a bit off as well.

**Comment by joegallo** [comment]
@elasticmachine update branch

**Comment by joegallo** [comment]
I'm pretty sure this is going to fail the test introduced by https://github.com/elastic/elasticsearch/pull/94281, which means that this PR is a bugfix for https://github.com/elastic/elasticsearch/issues/83653. /cc @HiDAl

**Comment by breskeby** [comment]
@elasticsearchmachine test this please

**Comment by HiDAl** [comment]
Yesterday I had to revert the test because it was failing due to warning headers. Today I pushed and merged PR #94388 which addresses the issue. Please rebase/merge on this PR @joegallo @felixbarny

**Comment by felixbarny** [comment]
I've updated to main and re-adjusted the test. I ran that test locally and I can confirm that it prevents the final pipeline from running twice.

**Comment by joegallo** [comment]
@elasticmachine update branch

**Review by joegallo — APPROVED** [review]

**Comment by felixbarny** [comment]
> I'm not especially enamored with the way this tracks indexRecursionDetection -- there's a somewhat similar executedPipelines that's being tracked by IngestDocument itself. Having IngestDocument track the index changes and reroutes like that would simplify the parameters to executePipelines, and I think it would make later changes like adding the index history a bit simpler.

I completely agree with that. Ideally, ingest pipelines and scripts should not be able to modify the history so that we can be certain whether a document is sent directly to a target or whether it has been rerouted.

> I'll take on moving indexRecursionDetection over to IngestDocument as a refactoring in a separate PR.

Thanks!

> Let's merge this and start iterating on https://github.com/elastic/elasticsearch/pull/76511.

Thanks for the review and the approval! I'll wait for @dakrone's approval before merging.

**Review by dakrone — APPROVED** [review]
LGTM also, thanks for all the work on this Felix & Joe

**Pull request merged**
